core/project_settings.h

```diff
@@ -218,4 +232,19 @@ Variant _GLOBAL_DEF_ALIAS(const String &p_var, const String &p_old_name, const V
 #define GLOBAL_DEF_ALIAS_RST(m_var, m_old_name, m_value) _GLOBAL_DEF(m_var, m_old_name, m_value, true)
 #define GLOBAL_GET(m_var) ProjectSettings::get_singleton()->get(m_var)
 
+/////////////////////////////////////////////////////////////////////////////////////////
+// Cached versions of GLOBAL_GET.
+// Cached but uses a typed variable for storage, this can be more efficient.
+#define GLOBAL_GET_CACHED(m_type, m_setting_name) ([](const char *p_name) -> const m_type & {\
+static_assert(std::is_trivially_destructible<m_type>::value, "GLOBAL_GET_CACHED must use a trivial type that allows static lifetime.");\
+static m_type local_var;\
+static uint32_t local_version = 0;\
+uint32_t new_version = ProjectSettings::get_singleton()->get_version();\
+if (local_version != new_version) {\
+	MutexLock lock(ProjectSettings::get_singleton()->get_global_mutex());\
+	local_version = new_version;\
+	local_var = ProjectSettings::get_singleton()->get(p_name);\
+}\
+return local_var; })(m_setting_name)
```

Comment:
I think we can make do without a lock:

```suggestion
#define GLOBAL_GET_CACHED(m_type, m_setting_name) ([](const char *p_name) -> const m_type & {\
static_assert(std::is_trivially_destructible<m_type>::value, "GLOBAL_GET_CACHED must use a trivial type that allows static lifetime.");\
static std::atomic<Pair<uint32_t, m_type>> cache;\
uint32_t new_version = ProjectSettings::get_singleton()->get_version();\
Pair<uint32_t, m_type> current_pair = cache.load(std::memory_order_acquire);
if (pair.first == new_version) {\
	return pair.second;\
}\
m_type new_var = ProjectSettings::get_singleton()->get(p_name);\
cache.store(Pair(new_version, new_var), std::memory_order_release);\
return new_var;\
})(m_setting_name)
```

From a quick test this is 3x faster than `std::mutex`, at least as long as there's few updates to the value.